### PR TITLE
[stable/prometheus] RBAC, namespaces, and node-exporter default update

### PR DESCRIPTION
**Pull request by alejandroEsc**
- rbac is off by default, but can be turned on
- added clusterroles, bindings and service accounts
- added namespaces where needed.
- updated default node-exporter because previous version had a fairly serious bug disallowing it to start up properly.

### CODE REVIEW

**Comment by alejandroEsc** [comment]
@mgoodness could you please review or assign?

**Comment by alejandroEsc** [comment]
@mgoodness could I get an eta on this review? if you are too busy any suggestions to possible maintainers that may be able to take over?

**Comment by cortesi** [comment]
@mgoodness I'd like to add my voice to @alejandroEsc here. The Prometheus chart is critical, and it is often broken for months on end with all PRs self-assigned to you and no activity.

File: stable/prometheus/templates/alertmanager-deployment.yaml
Comment (mgoodness) [inline stable/prometheus/templates/alertmanager-deployment.yaml]: Use `serviceAccountName: {{ if .Values.createRbac }}{{ template "prometheus.alertmanager.fullname" . }}{{ else }}"{{ .Values.alertmanager.serviceAccountName }}"{{ end }}`

File: stable/prometheus/templates/kube-state-metrics-deployment.yaml
Comment (mgoodness) [inline stable/prometheus/templates/kube-state-metrics-deployment.yaml]: Use `serviceAccountName: {{ if .Values.createRbac }}{{ template "prometheus.kubeStateMetrics.fullname" . }}{{ else }}"{{ .Values.kubeStateMetrics.serviceAccountName }}"{{ end }}`

File: stable/prometheus/templates/node-exporter-daemonset.yaml
Comment (mgoodness) [inline stable/prometheus/templates/node-exporter-daemonset.yaml]: Use `serviceAccountName: {{ if .Values.createRbac }}{{ template "prometheus.nodeExporter.fullname" . }}{{ else }}"{{ .Values.nodeExporter.serviceAccountName }}"{{ end }}`

File: stable/prometheus/templates/server-deployment.yaml
Comment (mgoodness) [inline stable/prometheus/templates/server-deployment.yaml]: Use `serviceAccountName: {{ if .Values.createRbac }}{{ template "prometheus.fullname" . }}{{ else }}"{{ .Values.server.serviceAccountName }}"{{ end }}`

File: stable/prometheus/templates/rbac-clusterrolebindings.yaml
Comment (mgoodness) [inline stable/prometheus/templates/rbac-clusterrolebindings.yaml]: Delete all YAML separators, and break the file up (one per CRB). Grouping resources like this can cause problems when upgrading.

File: stable/prometheus/templates/rbac-clusterrolebindings.yaml
Comment (mgoodness) [inline stable/prometheus/templates/rbac-clusterrolebindings.yaml]: Use `{{- if .Values.createRbac -}}`

File: stable/prometheus/templates/rbac-clusterrolebindings.yaml
Comment (mgoodness) [inline stable/prometheus/templates/rbac-clusterrolebindings.yaml]: We only support the last two versions of Kubernetes (currently 1.6 & 1.7), so it's safe to statically set the API version to `v1beta1`.

File: stable/prometheus/templates/rbac-clusterrolebindings.yaml
Comment (mgoodness) [inline stable/prometheus/templates/rbac-clusterrolebindings.yaml]: Use `name: {{ template "prometheus.fullname" . }}`

Also add standard labels (app, chart, release, component, heritage).

File: stable/prometheus/templates/rbac-clusterrolebindings.yaml
Comment (mgoodness) [inline stable/prometheus/templates/rbac-clusterrolebindings.yaml]: Use `name: {{ template "prometheus.fullname" . }}`

File: stable/prometheus/templates/rbac-clusterrolebindings.yaml
Comment (mgoodness) [inline stable/prometheus/templates/rbac-clusterrolebindings.yaml]: Use `name: {{ template "prometheus.fullname" . }}`

File: stable/prometheus/templates/rbac-clusterrolebindings.yaml
Comment (mgoodness) [inline stable/prometheus/templates/rbac-clusterrolebindings.yaml]: Delete

**Review by mgoodness — CHANGES_REQUESTED** [review]
The changes I'm requesting generally follow the RBAC pattern we established in #1504. I only commented on the first part of `rbac-clusterrolebindings`, but the changes should be applied throughout that file, `rbac-clusterroles`, and `rbac-serviceaccounts`.

**Comment by mgoodness** [comment]
Fixes #1075

**Comment by alejandroEsc** [comment]
@mgoodness thanks for the review, will start ASAP.

**Comment by alejandroEsc** [comment]
@mgoodness I believe I addressed all your points. The only thing I did not address was the apiVersion rbac option. You suggest I leave it as v1beta1 and remove this value. But I think until we are at v1 it still a good idea.

**Comment by naveensrinivasan** [comment]
@mgoodness Any update on this?

**Comment by chainlink** [comment]
Also waiting on this one to test prometheus on GCE

**Comment by maartenvanvliet** [comment]
I have enabled Legacy authorization on GCE to test prometheus. Works great. Waiting for this to be accepted though, and disable Legacy authorization again.

**Comment by chainlink** [comment]
Good to know @maartenvanvliet !

**Comment by mgoodness** [comment]
Reviewing right now. Sit tight.

**Review by mgoodness — APPROVED** [review]

**Comment by mgoodness** [comment]
serviceAccountNames now defaults to `default` if `rbac.create=false`, as is our pattern.

I also set the RBAC apiVersion to v1beta1. When it hits v1, we'll handle compatibility with a conditional based on the `.Capabilities` object.

Please test one more time to make sure I didn't bork anything.

**Comment by jurgenweber** [comment]
eta on merge? This would be really handy.

File: stable/prometheus/templates/alertmanager-clusterrolebinding.yaml
Comment (seh) [inline stable/prometheus/templates/alertmanager-clusterrolebinding.yaml:19]: - [ ] Given that Alertmanager only _reads_ objects, giving it the "cluster-admin" _ClusterRole_ is rather generous.  
  Better would be to define a _ClusterRole_ that allows it to read only what it needs to read.

File: stable/prometheus/templates/node-exporter-clusterrolebinding.yaml
Comment (seh) [inline stable/prometheus/templates/node-exporter-clusterrolebinding.yaml:19]: - [ ] Given that Node Exporter only reads things, giving it the "cluster-admin" _ClusterRole_ is rather generous.  
  Better would be to define a _ClusterRole_ that allows it to read only what it needs to read. What does it need to read from the API server?

File: stable/prometheus/values.yaml
Comment (seh) [inline stable/prometheus/values.yaml]: - [x] You risk granting the "cluster-admin" _ClusterRole_ to the "default" service account here.

Comment (mgoodness) [reply stable/prometheus/values.yaml]: Pods only run as `.Values.serviceAccountName` if `rbac.create=false`. In that case, none of the RBAC resources are created and it's up to the user to configure permissions externally.

Comment (seh) [reply stable/prometheus/values.yaml]: That comment above is confusing, then. It says, "when [RBAC] is enabled," which I took to mean "when this chart plays along with RBAC."

**Comment by mgoodness** [comment]
@jurgenweber I'd been [waiting](https://github.com/kubernetes/charts/pull/1443#issuecomment-322320396) for someone to deploy & test this before I merged. Also, like @seh, I'm somewhat uncomfortable running alertmanager & node exporter as `cluster-admin`. However, if the consensus is to improve RBAC in another PR, I'll get this merged today.

File: stable/prometheus/values.yaml
Comment (mgoodness) [reply stable/prometheus/values.yaml]: Fair enough. I'll be explicit about `rbac.create=false`.

**Comment by chainlink** [comment]
Tested this on gce with 1.6 and it works great out of the box!

**Comment by jurgenweber** [comment]
@mgoodness ok, I started giving it a go and then go distracted but as far as I could see all of the resources were green. Not a great test I know, but it is a fresh install for me.

File: stable/prometheus/templates/alertmanager-clusterrolebinding.yaml
Comment (seh) [reply stable/prometheus/templates/alertmanager-clusterrolebinding.yaml:19]: This still strikes me as dangerous. If an administrator installs this chart into a namespace in which other non-administrators can run pods, those other users can run a pod using this service account, and escalate their privileges up to "cluster-admin." I think you should at least document this hazard if you're not willing to come up with a safer, narrower _ClusterRole_ for Alertmanager here.

File: stable/prometheus/templates/node-exporter-clusterrolebinding.yaml
Comment (seh) [reply stable/prometheus/templates/node-exporter-clusterrolebinding.yaml:19]: [Likewise](https://github.com/kubernetes/charts/pull/1443/files#r133989161), but here with Node Exporter.

File: stable/prometheus/templates/alertmanager-clusterrolebinding.yaml
Comment (mgoodness) [reply stable/prometheus/templates/alertmanager-clusterrolebinding.yaml:19]: Any refinement of the the RBAC permissions can & should be submitted through PRs.

**Comment by jurgenweber** [comment]
thanks for the merge, for the record I upgraded to kube 1.7.4 and it works great. :) Thanks again.

**Pull request merged**
